File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.java
class Solution {
    int[] ans;
    Map<Integer, List<Integer>> adj;

    public int[] restoreArray(int[][] adjacentPairs) {
        // Initialize the result array and adjacency list
        ans = new int[adjacentPairs.length + 1];
        adj = new HashMap<>();

        // Populate the adjacency list based on the input pairs
        for (int[] a : adjacentPairs) {
            adj.putIfAbsent(a[0], new ArrayList<>());
            adj.putIfAbsent(a[1], new ArrayList<>());

            adj.get(a[0]).add(a[1]);
            adj.get(a[1]).add(a[0]);
        }

        // Find a node with only one neighbor to start the reconstruction
        for (Map.Entry<Integer, List<Integer>> e : adj.entrySet()) {
            if (e.getValue().size() == 1) {
                // Start DFS from the identified node
                dfs(e.getKey(), Integer.MIN_VALUE, 0);
                // Return the reconstructed array
                return ans;
            }
        }

        return ans;
    }

    // Depth-First Search to reconstruct the array
    private void dfs(int node, int prev, int ptr) {
        // Store the current node in the result array
        ans[ptr] = node;

        // Explore neighbors
        for (int nbr : adj.get(node)) {
            // Avoid revisiting the previous node
            if (nbr != prev) {
                // Recursively explore the neighbor
                dfs(nbr, node, ptr + 1);
            }
        }
    }
}

File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.py
class Solution:
    def restoreArray(self, adjacentPairs: List[List[int]]) -> List[int]:
        
        def bfs(node):
            ptr=0
            q=[(node, None)]
            
            while q:
                curr, prev = q.pop(0)
                ans[ptr] = curr
                ptr += 1
                for nbr in adj[curr]:
                    if nbr != prev:
                        q.append((nbr, curr))
        
        def dfs(node, prev=None, ptr=0):
            ans[ptr] = node
            for nbr in adj[node]:
                if nbr != prev:
                    dfs(nbr, node, ptr+1)
                
                
        
        adj = defaultdict(list)
        
        for a,b in adjacentPairs:
            adj[a].append(b)
            adj[b].append(a)
        ans = [0]*(len(adjacentPairs)+1)
        for k,v in adj.items():
            if len(v)==1:
                # bfs(k)
                dfs(k)
                return ans